Add support for F#-style Pipeline Operator

Pokute opened the pull request:
PR adding [pipeline operator](https://github.com/tc39/proposal-pipeline-operator) support to TypeScript. The proposal is currently a stage 1 proposal for EcmaScript. It "introduces a new operator |> similar to F#, OCaml, Elixir, Elm, Julia, Hack, and LiveScript, as well as UNIX pipes and Haskell's &. It's a backwards-compatible way of streamlining chained function calls in a readable, functional manner, and provides a practical alternative to extending built-in prototypes."

```
const square = (n: number) => Math.pow(n, 2);
4 |> square |> console.log; // 16
```

Currently this is a minimal implementation of pipeline operator, but I'm planning on making a F# variant support and after that a Topic/Smart pipeline variant PRs.

[Partial application](https://github.com/tc39/proposal-partial-application) is a good pair for minimal/F# pipeline operator and has it's own PR at #37973. I'm planning on making a separate draft PR that has both features in them.

There is also a [PR for Hack-style pipelines](https://github.com/microsoft/TypeScript/pull/43617) with `#` as a placeholder token.

## Code status

This works minimally. The TS transformer generates valid JS and there's working type inference.  There are many bugs, inconsistencies with specs, insufficient error messages currently. Most of failing tests are not included in this PR since Playground won't get generated with any failures and having faulty baselines is shady. I'm eagerly awaiting feedback, testing and even help with this PR. Even using and talking of these features will help advance them through TC39 proposal process!

`await` is currently not supported.

I used @rbuckton 's work on pipelineStage1 branch as a guide.

## Testing

[Playground link](https://www.staging-typescript.org/play?ts=4.3.0-pr-38305-46)


To test it in your project, add following to your package.json:
```
{
    "devDependencies": {
        "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/100059/artifacts?artifactName=tgz&fileId=D3836CF7DD5F6438481C5BCDBEE7A63AB21880164FEFD2BD75085B15EB974FCE02&fileName=/typescript-4.3.0-insiders.20210404.tgz"
    }
}
```
and then run `npm install`.

## PR state

* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [X] Code is up-to-date with the `master` branch
* [X] You've successfully run `gulp runtests` locally
* [X] There are new or updated unit tests validating the change

## Why a draft PR for stage 1 proposal?

The TypeScript officially follows EcmaScript / stage 3 proposals. However, I'm opening this draft pull request despite it being only stage 1. The partial application and pipeline proposals now are in a stage where they want people to test the features - write it in their code, find problems, test existing codebases. The usual way these would be done is through a Babel plugin, but since a few years ago, just a plain text editor with syntax highlighting is no longer the default JS development environment.

TypeScript is no longer just a build step. It's used by a significant portion of JS developers for all their development, even when prototyping new code. If the tool chain of these developers requires TypeScript, then a simple Babel plugin won't help. This means that writing TypeScript with those features is impossible without first-class support for them in TypeScript.

But then could we ask them to write just plain JS instead? It turns out that TypeScript is more and more part of normal JavaScript development process with Language Server Protocol pointing out errors in code and providing typing information. Having people encounter red squigglies and disappearing type information as soon as they use the new language feature will not count favourably to the user experience. I believe having TypeScript support for new syntax proposals as important or even more important than having a Babel plugin.

I was considering creating and upkeeping a separate fork, but that would just mean duplication of work. There's already a great infrastructure for TypeScript with playgrounds, CI, etc. With a draft PR, new features can more easily get more eyes, testers and feedback. There will eventually be issues and questions on new features for TypeScript and github.com/microsoft/TypeScript is where anyone would first look up info on partial application.

Of course, I'm not expecting merging until the proposal reaches stage 3.

orta commented:
@typescript-bot pack this.

markusjohnsson commented:
Nice! I have wondered myself how pipeline operators would play with typescript, especially how the types would flow. You could have went the way of only parsing and type checking, and letting e.g. babel do translation, but since you have also done translation this is really easy to try out. 

I did test it out and ran in to a type inference limitation which I hope can be fixed or it will be hard to use with complex types:

```ts
const ns = [1, 2, 3, 4, 5];

function map<T, R>(f: (t: T) => R) {
    return (ts: T[]): R[] => ts.map(f);
}

const result = ns |> map(x => x * x);
```

```
index.ts:8:16 - error TS2571: Object is of type 'unknown'.

8 ns |> map(x => x * x);
```

Here `T` must be inferred from the type of `ns`.

rbuckton commented:
@Pokute: If it helps at all I have two old branches where I started to look into implementing pipeline as well: 
- https://github.com/microsoft/TypeScript/tree/pipelineStage1 - Limited to only the pipeline operator
- https://github.com/microsoft/TypeScript/tree/rbuckton/pipeline - Pipeline, partial application, and a few other concepts. Older than pipelineStage1 and much more experimental

Pokute commented:
> Nice! I have wondered myself how pipeline operators would play with typescript, especially how the types would flow. You could have went the way of only parsing and type checking, and letting e.g. babel do translation, but since you have also done translation this is really easy to try out.

Writing the transformer was actually pretty simple as soon as I had the parsing and type checking in place.

> I did test it out and ran in to a type inference limitation which I hope can be fixed or it will be hard to use with complex types:
> Here T must be inferred from the type of ns.

I think that the form that TypeScript seems to internally handle it is `map(x => x * x)(ns)`.
This is an interesting case where the type of a function would be resolved from it's usage and it's not directly either. I'm interested whether there are existing cases of this happening within TypeScript.

In the simplest cases, it feels like it should be possible to resolve the type. Beyond that is out of my expertise. :-)

Either way, thank you for an interesting test case!

phaux commented:
Is there a way to try pipeline operator and partial application together?

dustinlacewell commented:
When I found this PR I nearly cried. Thank you.

lukewestby commented:
Hey folks, I stumbled upon this PR from https://github.com/tc39/proposal-pipeline-operator#build-tools. I'm not able to access the tarball mentioned here https://github.com/microsoft/TypeScript/pull/38305#issuecomment-623049381 and I'm just wondering if it's been removed intentionally.

Pokute commented:
> Hey folks, I stumbled upon this PR from https://github.com/tc39/proposal-pipeline-operator#build-tools. I'm not able to access the tarball mentioned here [#38305 (comment)](https://github.com/microsoft/TypeScript/pull/38305#issuecomment-623049381) and I'm just wondering if it's been removed intentionally.

At least I haven't removed the tarball. I wonder if those have limited lifetimes. Any ideas @orta?

orta commented:
Yep, looks like they timeout - we're having a chat about what to do in the long term, but for now I'll give it another 10-12 days.

@typescript-bot pack this

mikearnaldi commented:
> Nice! I have wondered myself how pipeline operators would play with typescript, especially how the types would flow. You could have went the way of only parsing and type checking, and letting e.g. babel do translation, but since you have also done translation this is really easy to try out.
> 
> I did test it out and ran in to a type inference limitation which I hope can be fixed or it will be hard to use with complex types:
> 
> ```ts
> const ns = [1, 2, 3, 4, 5];
> 
> function map<T, R>(f: (t: T) => R) {
>     return (ts: T[]): R[] => ts.map(f);
> }
> 
> const result = ns |> map(x => x * x);
> ```
> 
> ```
> index.ts:8:16 - error TS2571: Object is of type 'unknown'.
> 
> 8 ns |> map(x => x * x);
> ```
> 
> Here `T` must be inferred from the type of `ns`.

Running in the same issue, without this sort of inference the usage of the operator is very limited, not sure if this can help but currently the only place I sow proper inference with a pipe implementation is this (not the operator a simple pipe function):
https://github.com/gcanti/fp-ts/blob/master/src/pipeable.ts#L97

xixixao commented:
This is great, I might take a stab at adding the Hack-style proposal (see https://github.com/tc39/proposal-pipeline-operator/issues/167#issuecomment-631878219 - it's the Smart proposal without the minimal proposal) to Flow.

stken2050 commented:
One of the greatest updates.
Just in case, here's the working:

```
 "devDependencies": {
    "@types/node": "*",
    "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/74014/artifacts?artifactName=tgz&fileId=53B4D99407A93B4AF6FAEF8437B02126A8C7FEC62F1C3307DAA098F75D2CE5D602&fileName=/typescript-4.0.0-insiders.20200515.tgz"
  }
```

```
 const log = msg => (console.log(
    typeof msg === "function"
      ? msg
      : JSON.stringify(msg))
    , msg);

```

![image](https://user-images.githubusercontent.com/1316994/85087525-b8f14400-b218-11ea-8cdf-347db00466a7.png)

sgioia9 commented:
YESSSS

chenglou commented:
Funny that this is coming back full circle. I've made a comment a while ago regarding the type inference/tooling problem with the pipe, and how to solve it: https://github.com/tc39/proposal-pipeline-operator/issues/143#issuecomment-492019845

Jopie64 commented:
About that backward type inference problem:

In RxJS, TypeScript works just fine for a situation like this:

```javascript
const int$ = of('hello').pipe(map(x => x.length));
// type of int$ is inferred to be Observable<number>
```

This would be equivalent to:

```javascript
const int$ = of('hello') |> map(x => x.length);
```

Why would it be impossible to infer `int$` in the latter case while it is inferrable in the former? In both cases `map` is something of type:
```javascript
type Map = <T, U>(f: T => U) => Observable<T> => Observable<U>
```
And hence the receiver is the last curried argument. So it feels like it should be inferrable in both cases..?

benlesh commented:
FYI: It appears the type inference isn't working that great for more advanced cases like RxJS [see this playground example](https://www.typescriptlang.org/play/index.html?ts=4.0.0-pr-38305-12#code/JYOwLgpgTgZghgYwgAgMoFcBGBnBVgAOYwA9iMgN4BQyy6I2Wu+mEAFAJQBcyAbicAAmAbioBfKlVCRYiFAHkc0XtAA8AFQB8lGshAQAHmDa84AG3QQe67nwEjd0KCShsnPOCACet-kNG0CCQAtgRmEJCcPH4OElLg0PBIyIrYynCY4Rra1LSMOHjArGwkSlAqUDwACnBQxOaqqcpqWpq2GAX4RKQgonFgXgQKBBoANMgAStoAvMhs2CToUEg8TeUZWVocyNPaa6aZEKpTolSCEAhmtSgw9AjEZMjBcCPq41NsMCA8JuaW1uNQOcDDwQOhgqwoNtdpNbPJXu9NKdzpdrshbiB7j10cAzDJsp9vnNTBYrMg3sggYZQeDIdDtJgSCRwp44QjyUjJCirlAUEEGGBkAslisUmUDlkwRDoJyqPzsILeYw8TshYtlhBdAAfbTPAhsAw7bSGgDUyAMHG1upeBqN5uQZoA5ABCV2Oy20HU4vHQW0wgwAOnCIAA5mAABbIVTIABMAAYOMIgA)

Copied here for convenience:

```ts
interface Subscription {
  unsubscribe(): void;
}

interface Observer<T> {
  next(value: T): void;
  error(err: any): void;
  complete(): void;
}

interface Observable<T> {
  subscribe(observer: Partial<Observer<T>>): Subscription;
}

type Op<T, R> = (source: Observable<T>) => Observable<R>;

declare function map<T, R>(fn: (value: T, index: number) => R): Op<T, R>;

declare function filter<T>(fn: (value: T, index: number) => boolean): Op<T, T>;

declare const source: Observable<number>;

const result = source
  |> map(x => x + x) // <-- x is `unknown`
  |> map(x => x + '!!!') // <-- x is `unknown`
  |> filter(x => x.length < 20); // <-- x is `unknown`
```

This is a real shame, because it would be a killer language feature for libraries like RxJS.

lozandier commented:
@benlesh @markusjohnsson Seems to be a pattern of type inference needing to be added for its use w/ generics; generic use cases don't seem to be covered yet in `cases/conformance` nor `baselines/reference` yet by @Pokute.

stken2050 commented:
@typescript-bot pack this

orta commented:
@typescript-bot pack this

kiprasmel commented:
Testing this out in a project and having to manually update the packed bundle because it expires in about 10 days is very annoying - could we get a feature to tell typescript bot to pack this indefinitely? Thanks!

/cc @orta

kasperpeulen commented:
Actually, this PR convinces me that the pipeline operator in its current form is not a good match for typescript.
The pipeline operator feels the most natural if we live in a js world with curried, data last, functions.
Of course, Haskell and other FP languages have moved to this pattern for similar reasons. It makes function much easier composable.

However:
1. This is not the current js world where we live in. I think any javascript programmer with no FP background, will use the data as the first argument. This is the standard for most JS programmers, codebases and I also think still the standard in most open source APIs. Do we really want to change this whole paradigm? 
2. Curried functions are quite a bit uglier in js, especially if you want to use `function`, `function*` or `async function`.
3. If you are only going to use the curried function by itself, it looks more confusing. For example, what is the output of this function: `concat([1])([2, 3])`? `[1,2,3]` or `[2,3,1]`
4. TS can not really infer curried, data last, functions well. This has nothing to do with this PR, but more in general that TS infers from left to right, (data first), and not right to left (data last).

For example, this also get not inferred well:

```ts
interface Observable<T> {}
type Op<T, R> = (source: Observable<T>) => Observable<R>;
declare function map<T, R>(fn: (value: T, index: number) => R): Op<T, R>;
declare function filter<T>(fn: (value: T, index: number) => boolean): Op<T, T>;
declare const source: Observable<number>;
const result = filter(x => x.length < 20)(map(x => x + '!!!')(map(x => x + x)(source)));
```

Libraries like RxJS partially has solved this problem, by having a sick pipe method, that basically allows curried functions to be inferred left to right again. For example, it could be typed like this:

```ts
export function pipe<X1, X2, X3, R>(
  value: X1,
  f1: (x: X1) => X2,
  f2: (x: X2) => X3,
  f3: (x: X3) => R,
): R;

const result = pipe(
  source,
  map(x => x + x),
  map(x => x + '!!!'),
  filter(x => x.length < 20),
);
```

But if typescript can not infer curried functions well using the pipeline "natively", TS users will still prefer a custom pipe method above the js operator. 

It sounds like me, that we may need to reach the same conclusion as rescript did (used to be called reasonml), and move to a pipeline operator that is optimized to work with data first functions:
https://www.javierchavarri.com/data-first-and-data-last-a-comparison/

lozandier commented:
@kasperpeulen Your mention of RxJS in your last comment is kind of confusing. **The main contributors of RxJS have also campaigned for this version pipeline operator for *some* time; the version of pipeline operator they've supported that this PR is related to is in fact a drop-in replacement for their implementation.**

A good example of that can be found here: https://reactive.how/rxjs/explorer

Accordingly it *can* be inferred just fine by TypeScript; it been pointed out by people like @benlesh that this current attempt of realizing that version is not acting as expected well within the capabilities of TypeScript today.

Similarly, pipeline operator is also a drop-in replacement for the equivalent in `lodash`, `ramda`, & other popular functional programming libraries that also have significant traction in the JS community overall. The way this variant of the pipeline operator works also aligns with how POSIX-compliant unix shells work & so on. **Overall, how this pipeline operator behaves is not an uncommon way data pipelining works**.  

**Accordingly, I respectfully disagree we need to prioritize an alternate pipeline operator that "works on data first functions" that isn't as idiomatic or have been historically leveraged by functional programmers  ; not opposed to that being pursued as a separate initiative/standard proposed nonetheless.**

kasperpeulen commented:
@lozandier My point is that the curried function of Rx.js only infer in combination with the pipe method. They don't infer if you would use the curried function directly:

`const result = filter(x => x.length < 20)(map(x => x + '!!!')(map(x => x + x)(observable)));`

That is I think the reason why the pipeline operator implemented in this PR, can not infer them. 
It could be, that we can change the pipeline operator implementation in TS so that this will get inferred as good as the `.pipe` method of RxJS.
However, isn't a bit silly that we are about to change the whole paradigm of how to write a function in javascript, while we can only leverage those functions using a pipeline operator, and not be able to write those functions directly?

lozandier commented:
@kasperpeulen I’m not sure what you mean by “change the paradigm of how functions are written in JS” when partial application capabilities & the pipeline operator enable the use of existing functions that JS developers have written just fine.

This is particularly true of *referential transparent functions* that have been augmented by the equivalent of both a variety of FP libraries provide FP devs today while they wait for these two proposals to be supported by the language properly usually  through `ramda`, `lodash`, or from scratch via `call`/`apply`.

These functions are extremely common, particularly when writing optimally testable JS anyway; I’m not sure what functions would have to be majorly re-written for use w/ the pipeline operator when combined w/ the partial application proposal.

The pipeline operator on its own elegantly supports single arg (unary) referential transparent functions while combined w/ the partial application proposal enables it to handle all typical functions JS write today just fine.

Without the latter being available to the language for a long time is where I can see your concern having merit forcing *point-free programming* all the time when using the pipeline operator; that said, all steps point to the pipeline operator & partial application capabilities being combined or ratified simultaneously to alleviate your concern. 

Currently the [F#-style](https://github.com/tc39/proposal-pipeline-operator/wiki#proposal-1-f-sharp-style-only-with-await), [split mix](https://github.com/tc39/proposal-pipeline-operator/wiki#proposal-3-split-mix), & [smart mix](https://github.com/tc39/proposal-pipeline-operator/wiki#proposal-4-smart-mix) are proposals that explicitly do so. 

The latter 2 are explicitly a florian's compromise of accepting some of the ideas of a recently proposed new style (Hack-style) from a perspective similar to yours being cautious of forcing point-free programming w/ the use of pipeline-operator.

Split mix & F#-style I find promising on that matter.

kasperpeulen commented:
With the partial application proposal also in, you indeed dont have to rewrite existing javascript, but it is still more elegant to use the curried variant of the same function (to avoid the ?).

I guess the js ecosystem will move more to curried, data last functions even with the partial application syntax.

The question is, do we want that? 

When data last, curried functions, when called directly as shown above, infer very poorly the generic parameters, which seems to be a core design limitation of typescript.

Or are there better solution for function composition in typescript?

lozandier commented:
@kasperpeulen While I prefer point-free programming, it's purely subjective that a curried variant of the same function is "still more elegant". 

As far as the two capabilities "moving the JS ecosystem" in a particular direction, that's up for the JS community to decide; not exactly sure who is "we" in your original comment & why their wants should outweigh the JS community.

The existence of these proposals won't force the JS community in a particular direction; these capabilities being prevalent in codebases after being ratified would merely signal & verify people wanted to go this direction and them being available natively finally more easily enabled that. The existence of the pipeline operator & partial application proposal natively unblocks barrier of entry selling & communicating longly-leveraged patterns of writing declarative, functional code involving left-to-right composition, clearer mix-in patterns, & so on in environments where every byte counts compared to importing & writing ad hoc code to approximate their behavior with limitations a native solution doesn't. 

 With these capabilities in the language soon, it will be marketed that there are "new" ways to create declarative, functional code in JS. 

As far as limitations w/ TypeScript's type system: TypeScript core maintainers are a brilliant bunch; at the end of the day JS doesn't revolve around TypeScript; as a superset of JS they'll refactor TypeScript as needed to continue to be that in good faith. 

It can be argued w/ the pipeline operator types flow much easier, not the other way around.

kasperpeulen commented:
> With these capabilities in the language soon, it will be marketed that there are "new" ways to write declarative, functional code in JS as a result. That's fine.

That is fine, if you believe this is the best way to write declarative functional code in JS using curried data last functions. That is what I doubt. 

Your argument are not really argumenting why data last curried functions with say a F# pipeline are a better fit for javascript, than say a data first “normal” function with an elixir like pipeline. 

This article goes in detail about the pros and cons about those approaches:
https://www.javierchavarri.com/data-first-and-data-last-a-comparison/

Probably better to move this discussion to the pipeline repo though. 

If we indeed can fix the all typescript inference issues for this PR and curried functions in general, than it wouldnt matter that much to me which variant of the pipeline proposal is chosen.

But if those issues are too hard to fix, it may be another argument to take the elixer pipeline proposal more serious.

stken2050 commented:
> Heya @orta, I've started to run the tarball bundle task on this PR at [19fd8b1](https://github.com/microsoft/TypeScript/commit/19fd8b187e5acad445a204e17f0f3f1f285a10a6). You can monitor the build [here](https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_build/results?buildId=82439).

@orta The pack build seems failed?

stken2050 commented:
@orta @Pokute @kiprasmel  

I have the packed typescript-4.0.0-insiders.20200515.tgz, and now the latest version is  TypeScript v4.1.2 (Stable).
https://github.com/microsoft/TypeScript/releases

How can we obtain the repacked latest version with Pipeline Operator?

Thanks.

orta commented:
You're welcome to rebase this PR to get it running on 4.1 and we can re-issue a new build - I've also added a feature to the pack command so that a permanent npm deploy will be created the next time this is built https://github.com/microsoft/TypeScript-Make-Monaco-Builds/commit/92b5625f22bd29e43a9056714e84732f973eb08a

stken2050 commented:
@orta I really appreciate your contribution!

I've investigated 

https://github.com/microsoft/TypeScript-Make-Monaco-Builds/commit/92b5625f22bd29e43a9056714e84732f973eb08a

However, the things are not clear to me, so please explain more.

I would like to confirm which URL a permanent npm deploy will be uploaded.
According to
https://github.com/microsoft/TypeScript-Make-Monaco-Builds#cdnd
I found:
Releases: https://typescript.azureedge.net/indexes/releases.json
Pre-Releases: https://typescript.azureedge.net/indexes/pre-releases.json
Next: (provides a redirect to the latest nightly version)
https://typescript.azureedge.net/indexes/next.json

For instance my current package.json for my own project is

```
 "devDependencies": { 
    "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/74014/artifacts?artifactName=tgz&fileId=53B4D99407A93B4AF6FAEF8437B02126A8C7FEC62F1C3307DAA098F75D2CE5D602&fileName=/typescript-4.0.0-insiders.20200515.tgz"
  }
```

How can we adapt to a permanent npm deploy with repacked Pipeline Operator??

Thanks again.

orta commented:
The bot would give instructions after the new version has been built, https://github.com/microsoft/TypeScript-Make-Monaco-Builds/commit/225742ada04fed90b98112997566169be9f61ac0 - until this has been rebased, and re-packed, it won't show up [here](https://www.npmjs.com/package/@typescript-deploys/pr-build).

stken2050 commented:
> The bot would give instructions after the new version has been built, [microsoft/TypeScript-Make-Monaco-Builds@225742a](https://github.com/microsoft/TypeScript-Make-Monaco-Builds/commit/225742ada04fed90b98112997566169be9f61ac0) - until this has been rebased, and re-packed, it won't show up [here](https://www.npmjs.com/package/@typescript-deploys/pr-build).

@orta Is new version with pipeline-operator released?

I don't see any bot activity here or links...  Thanks.

orta commented:
This PR would need rebasing, and to be re-requested to be packaged again to get the new infra I added for permanent deploys

Lonli-Lokli commented:
@Pokute Can you please rebase it?

Pokute commented:
I rebased the code now.

Additionally I fixed a bunch of bugs - all typescript code inside pipelines after `|>` was not converted into JS. The tests now fail due to additional testing constrains added since last year, specifically it seems to complain about unused member field, `typeArguments`, in `PipelineExpression`. I think `typeArguments` will be used later so I left them in the code.

On the topic of type arguments, I actually tried to fix that one but it was more complex than I initially expected. If I went for `[] |> Array.from<number>;` syntax then I have to make a separate, temporary expression that's similar to a CallExpression, but the error reporting would become complex. A simpler way would be to use the funny-looking `[] |><number> Array.from;` which seems easier to implement. I will probably try the latter first since it would allow me to test the feature.

I'm not sure if this can be packed with existing errors, so I'll try to do some kind of fixes soon.

orta commented:
@typescript-bot pack this

stken2050 commented:
> I rebased the code now.
> 
> Additionally I fixed a bunch of bugs - all typescript code inside pipelines after `|>` was not converted into JS. The tests now fail due to additional testing constrains added since last year, specifically it seems to complain about unused member field, `typeArguments`, in `PipelineExpression`. I think `typeArguments` will be used later so I left them in the code.
> 
> On the topic of type arguments, I actually tried to fix that one but it was more complex than I initially expected. If I went for `[] |> Array.from<number>;` syntax then I have to make a separate, temporary expression that's similar to a CallExpression, but the error reporting would become complex. A simpler way would be to use the funny-looking `[] |><number> Array.from;` which seems easier to implement. I will probably try the latter first since it would allow me to test the feature.
> 
> I'm not sure if this can be packed with existing errors, so I'll try to do some kind of fixes soon.

Thanks a lot for your contribution!
Waiting for an update!

orta commented:
@typescript-bot pack this

zakhenry commented:
This is cool! Here's a contrived example of an interesting use-case for using pipelines to create builders composed from separate functions.

```ts
const url = new URLBuilder()
  |> host('localhost')
  |> port(4200)
  |> query('search', 'typescript')
  |> query('sort', 'recent')
  |> build
```

https://www.staging-typescript.org/play?ts=4.3.0-pr-38305-46#code/MYGwhgzhAECqBKAZAQgVwJYgCYFMBO0A3gFDTQAOqARiOsNAI6r4CeAXNAIJ55gsA8AbQgAXPOgB2AcwA00UeOkBdAHzQAvNEFKA3KQrVa9ABYB7UQH4OCyVL1lKNOhVN4RV6BNQBbKvj0AvsTEwKYSotBmEZoAFFEi1mK2AJQaKjGoeCBomLh4HAgoGNj4qepqJGSZ2cV5AHTxGpHmIvp4OCKZEtDVOSV4xEEhYRFMrE0xY3gsANI47PJJ0nJTLABqYCDMiYpSZWoZWX15BUjHpWlE+r21+HWrdZQQxjGCq3MsK8zTG1s4Ssk2h0uj0jrcBkNQuERC43BNyK4Ep4fH48PtoIcarl8Kciti0ZdKqCsf1Hoimgi3EDOnhujd8YNglCIlRahN6f1cecCeVoAADAAkhA59XiATYQpFd0pIgCFklYPx92+LDq3jA5EmlwYggADEpoABqaAAcnUJqNjEEAEYAXUAFamSQxE0AMhNyQCfKZIxh1SaEhwAHc4GdwTFAWQAD5qeIukCmYCbeIe-Qx2EiGIAFgATLrdZHoOnVi6IDgwHhgMYTXITSIWOQcBBgOJyCJU9G1CWTRBETXTe1gDgJO3C+nWbkfeFTCAcHUE1JMckgA

I can see this being very useful for working with external apis etc as you could easily add your own builder method without having to extend the external builder class. composition ftw!


Unfortunately [from the TC39 meeting notes](https://github.com/tc39/notes/blob/master/meetings/2021-03/mar-10.md#pipeline-operator) it seems like the pipeline proposal is starting to veer into the Hack style preference. Hopefully they can find a champion to make the case for the IMO simpler implementation demonstrated here. My three biggest use cases for pipeline (rxjs, array manipulation and builders) all would be less readable with Hack style :(

Pokute commented:
> Unfortunately from the TC39 meeting notes it seems like the pipeline proposal is starting to veer into the Hack style preference.

I'm sure that after we get any version of pipeline operator in place, it should be very straightforward to augment that functionality with another operator. For hack-style pipeline, this could be `#|>` or `|>>` or whatever to call unary functions.

Thus this all discussion now feels to be about preference and priorities and few are actually opposed to a more full set of features. The issue on priorities and preferences has already stalled this for multiple years and I recall It's only a very small minority that would rather see no pipelines rather than something other than _their_ version of pipelines. That's why I'm now supporting **any** pipeline variant - this one just happened to be the one I worked on first.

I'll probably try to implement a hack-style PR next instead of going for `typeArguments` -support.

btoo commented:
> Hopefully they can find a champion

according to the notes @DanielRosenwasser and his team have "some light interest in potentially championing this proposal"

this needs to happen yesterday 🤞

lozandier commented:
> This is cool! Here's a contrived example of an interesting use-case for using pipelines to create builders composed from separate functions.
> 
> ```ts
> const url = new URLBuilder()
>   |> host('localhost')
>   |> port(4200)
>   |> query('search', 'typescript')
>   |> query('sort', 'recent')
>   |> build
> ```
> 
> https://www.staging-typescript.org/play?ts=4.3.0-pr-38305-46#code/MYGwhgzhAECqBKAZAQgVwJYgCYFMBO0A3gFDTQAOqARiOsNAI6r4CeAXNAIJ55gsA8AbQgAXPOgB2AcwA00UeOkBdAHzQAvNEFKA3KQrVa9ABYB7UQH4OCyVL1lKNOhVN4RV6BNQBbKvj0AvsTEwKYSotBmEZoAFFEi1mK2AJQaKjGoeCBomLh4HAgoGNj4qepqJGSZ2cV5AHTxGpHmIvp4OCKZEtDVOSV4xEEhYRFMrE0xY3gsANI47PJJ0nJTLABqYCDMiYpSZWoZWX15BUjHpWlE+r21+HWrdZQQxjGCq3MsK8zTG1s4Ssk2h0uj0jrcBkNQuERC43BNyK4Ep4fH48PtoIcarl8Kciti0ZdKqCsf1Hoimgi3EDOnhujd8YNglCIlRahN6f1cecCeVoAADAAkhA59XiATYQpFd0pIgCFklYPx92+LDq3jA5EmlwYggADEpoABqaAAcnUJqNjEEAEYAXUAFamSQxE0AMhNyQCfKZIxh1SaEhwAHc4GdwTFAWQAD5qeIukCmYCbeIe-Qx2EiGIAFgATLrdZHoOnVi6IDgwHhgMYTXITSIWOQcBBgOJyCJU9G1CWTRBETXTe1gDgJO3C+nWbkfeFTCAcHUE1JMckgA
> 
> I can see this being very useful for working with external apis etc as you could easily add your own builder method without having to extend the external builder class. composition ftw!
> 
> Unfortunately [from the TC39 meeting notes](https://github.com/tc39/notes/blob/master/meetings/2021-03/mar-10.md#pipeline-operator) it seems like the pipeline proposal is starting to veer into the Hack style preference. Hopefully they can find a champion to make the case for the IMO simpler implementation demonstrated here. My three biggest use cases for pipeline (rxjs, array manipulation and builders) all would be less readable with Hack style :(

It's been abundantly clear in an overwhelming amount of threads with JS community participation that hack-style hasn't had much traction over the other styles. Accordingly, it would be hasty to prioritize it over the other styles. 

Simple proposal is a base that is supported without contention; accordingly I'm of the opinion that should be the implemented first & foremost.

lozandier commented:
> > Unfortunately from the TC39 meeting notes it seems like the pipeline proposal is starting to veer into the Hack style preference.
> 
> I'm sure that after we get any version of pipeline operator in place, it should be very straightforward to augment that functionality with another operator. For hack-style pipeline, this could be `#|>` or `|>>` or whatever to call unary functions.
> 
> Thus this all discussion now feels to be about preference and priorities and few are actually opposed to a more full set of features. The issue on priorities and preferences has seemingly already stalled this proposal for multiple years and I recall It's only a very small minority that would rather see no pipelines rather than something other than _their_ version of pipelines. That's why I'm now supporting **any** pipeline variant - this one just happened to be the one I worked on first.
> 
> I'll probably try to implement a hack-style PR next instead of going for `typeArguments` -support.

Wouldn't it be simpler to get the base characters working first prior to hack-style? Hack-style has been proposed as a florian's compromise with heavy pushback from the JS community, especially the functional programming community.

lozandier commented:
Hope this helps @zakhenry & @Pokute

orta commented:
Alright folks, let's not use the TypeScript issues to debate the merits of the F# vs Hack proposal - you can use [the proposal's repo](https://github.com/tc39/proposal-pipeline-operator) or twitter etc. Even better, if you work at a place which is a TC39 member, attend the next meeting and give feedback.

Pokute commented:
**Update**: I've been refactoring the TS PRs for pipelines (both F# #38305  and hack #43617 ) during this week and the commits are a lot cleaner now. The current commit history for hack-style pipelines is just a mess since it builds upon the F# implementation. The newer one will use a common commit.

One big idea I came upon is to have different PipelineExpression types for both a method that itself calls a unary function (F#) and another (hack) that just replaces tokens with the "argument". After all, the way those two pipelines work is different despite the similarities. This will allow for easier support of both F#, hack and even smart pipeline styles at the same time.

I believe that whichever style of F# or hack pipelines will be the default, the next shop would be to add another pipelineish operator for the other one. This is relatively straightforward to implement in TypeScript, so I'm considering adding [Tacit Unary Function Application](https://github.com/js-choi/proposal-hack-pipes#tacit-unary-function-application) syntax for this Hack-style PR to enable `Math.random() |> 1 / # |>> Math.round;`. For F# style pipelines, partial function application might be a better pair, so I won't be adding additional hack/topic-support to this PR unless there's a lot of requests for it. 

I've been rebasing my own repos against master, but I still haven't had time to fix the `(Missing)` errors in Hack-style pipelines. Hopefully I get new versions out before the end of the week.

orta commented:
@typescript-bot pack this

Lonli-Lokli commented:
Will it be possible to use pipeline with iterators, sync and async?

lozandier commented:
@Lonli-Lokli Yes, [see the following examples from the proposal's github repo](https://github.com/tc39/proposal-pipeline-operator#motivating-examples)

The simple pipeline handles the former 2 trivially by design; F#, Smart, & Hack-style proposals build on top of the simple proposal to accommodate async functions / promises / iterables a variety of ways w/ pros & cons.

rjhilgefort commented:
Is anyone aware of a working version of this PR? I tried the playground link posted by the bot, but the playground doesn't appear to be working with this branch?

https://www.staging-typescript.org/play?ts=4.3.0-pr-38305-57

<img width="1421" alt="image" src="https://user-images.githubusercontent.com/2034823/120235826-e3b98a00-c228-11eb-80a9-75f7a437169c.png">

Pokute commented:
> Is anyone aware of a working version of this PR? I tried the playground link posted by the bot, but the playground doesn't appear to be working with this branch?

There might be a problem with operator precedence. You need to wrap the inline arrow function within parenthesis:  `|> ((x: string) => x.toUpperCase()) |>`

Thank you for testing this out.

rjhilgefort commented:
@Pokute ah, thanks, I didn't realize you had to wrap the lamda in parens. That seems to work, mostly. The compiler accepts the expression, but still has an error.

![image](https://user-images.githubusercontent.com/2034823/120495273-c811c900-c379-11eb-89b0-7227fa104146.png)

It doesn't seem to properly infer that `x` is a string and I would have to tell the compiler that it's a string explicitly.

![image](https://user-images.githubusercontent.com/2034823/120495378-e1b31080-c379-11eb-92be-53c1d621e442.png)

Is that expected for this build? Is the plan to have that inference work?

stken2050 commented:
> @Pokute ah, thanks, I didn't realize you had to wrap the lamda in parens. That seems to work, mostly. The compiler accepts the expression, but still has an error.
> 
> ![image](https://user-images.githubusercontent.com/2034823/120495273-c811c900-c379-11eb-89b0-7227fa104146.png)
> 
> It doesn't seem to properly infer that `x` is a string and I would have to tell the compiler that it's a string explicitly.
> 
> ![image](https://user-images.githubusercontent.com/2034823/120495378-e1b31080-c379-11eb-92be-53c1d621e442.png)
> 
> Is that expected for this build? Is the plan to have that inference work?

I think it's natural that the pipeline operator has a high precedence.

and the error is common in typescript, nothing to do with with to the pipeline operator.

Pokute commented:
> > The compiler [...] still has an error.
> > It doesn't seem to properly infer that `x` is a string and I would have to tell the compiler that it's a string explicitly.
> > ![image](https://user-images.githubusercontent.com/2034823/120495378-e1b31080-c379-11eb-92be-53c1d621e442.png)
> > Is that expected for this build? Is the plan to have that inference work?
> 
> and the error is common in typescript, nothing to do with with to the pipeline operator.

<img width="286" alt="Screenshot 2021-06-02 at 18 21 05" src="https://user-images.githubusercontent.com/5271115/120507540-91669d80-c3cf-11eb-8023-a5ea2628f24c.png">

This should be inferable as shown in a above screenshot. It does require more work though, but I'm not sure when I can work on it.

I'm not sure about the precedence either. I'll need to check it out.

timbuckley commented:
TC39 is moving forward with the hack-proposal version, they just moved it to stage 2. Therefore, this PR should be closed in favor of https://github.com/microsoft/TypeScript/pull/43617

stken2050 commented:
> TC39 is [moving forward with the Hack-proposal version](https://github.com/tc39/proposal-pipeline-operator) (stage 1 -> 2).
> 
> Therefore, this PR should be closed in favor of the Hack-style approach #43617

I firstly thought it's good news, but investigating "hack-style" that for some reason I didn't notice (I only knew miminal/F#/Smart).

I think if a programmer knows mathematics or functional programming, the one will notice it's just a mess and night-mare.

I wrote this:

https://gist.github.com/tabatkins/1261b108b9e6cdab5ad5df4b8021bcb5#gistcomment-3892109

stken2050 commented:
I've just posted
https://github.com/tc39/proposal-pipeline-operator/issues/202#issuecomment-921433351
Hack-style contains serious problems.
I wish TypeScript users join and I would like to read your insight. Thanks.

stken2050 commented:
This is a new issue to discuss 
Impact of Hack pipes on functional programming
https://github.com/tc39/proposal-pipeline-operator/issues/217

stken2050 commented:
> This is a new issue to discuss
> Impact of Hack pipes on functional programming
> [tc39/proposal-pipeline-operator#217](https://github.com/tc39/proposal-pipeline-operator/issues/217)

I had to move the place, thanks.

https://github.com/tc39/proposal-pipeline-operator/issues/223

stken2050 commented:
**Pipelines with operator overloading**
https://github.com/tc39/proposal-pipeline-operator/issues/228

We don't want pipe any more. It's redundant over the operator overloading. We no longer support this pipeline-operator proposal as a whole.

RxJS author has just announced he no longer supports the Hack and will ignore.
https://github.com/tc39/proposal-pipeline-operator/issues/228#issuecomment-925465598

stken2050 commented:
I made an issue:
**Inconsistency of type inference of Hack-style pipeline-operator proposal (Stage2) #46101**

andrewbranch commented:
@stken2050 please do not post any comments here that are unrelated to the specific implementation of this PR. General discussion of the pipeline proposal should be done on the pipeline proposal’s repository. Seeing as you’re already participating there, one backlink for visibility was ok, but at this point you’re bordering on spam.

sandersn commented:
This experiment is pretty old, so I'm going to close it to reduce the number of open PRs.
